Security: don't call prepare index for reads

jaymode opened the pull request:
The security native stores follow a pattern where
`SecurityIndexManager#prepareIndexIfNeededThenExecute` wraps most calls
made for the security index. The reasoning behind this was to check if
the security index had been upgraded to the latest version in a
consistent manner. However, this has the potential side effect that a
read will trigger the creation of the security index or an updating of
its mappings, which can lead to issues such as failures due to put
mapping requests timing out even though we might have been able to read
from the index and get the data necessary.

This change introduces a new method, `checkIndexVersionThenExecute`,
that provides the consistent checking of the security index to make
sure it has been upgraded. That is the only check that this method
performs prior to running the passed in operation, which removes the
possible triggering of index creation and mapping updates for reads.

Additionally, areas where we do reads now check the availability of the
security index and can short circuit requests. Availability in this
context means that the index exists and all primaries are active.

Relates #33205

elasticmachine commented:
Pinging @elastic/es-security

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:
```diff
@@ -926,7 +938,7 @@ private void checkIfTokenIsRevoked(UserToken userToken, ActionListener<UserToken
             // index doesn't exist so the token is considered valid.
```
For this one we need indexExists. If the index doesn't exist then we make the assumption that the token has never been invalidated. If the index is not available, we cannot make the same assumption.

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:
```diff
@@ -673,30 +678,36 @@ private void findTokenFromRefreshToken(String refreshToken, ActionListener<Tuple
+                                attemptCount.incrementAndGet();
+                                findTokenFromRefreshToken(refreshToken, listener, attemptCount);
+                            } else if (searchResponse.getHits().getHits().length < 1) {
+                                logger.info("could not find token document with refresh_token [{}]", refreshToken);
```
This is not changed by this PR so I am leaving it as is.

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:
```diff
@@ -253,6 +253,22 @@ private static Version readMappingVersion(String indexName, MappingMetaData mapp
+     */
+    public void checkIndexVersionThenExecute(final Consumer<Exception> consumer, final Runnable andThen) {
+        final State indexState = this.indexState; // use a local copy so all checks execute against the same state!
+        if (indexState.indexExists && indexState.isIndexUpToDate == false) {
```
I added it to the docs. I do not think we should check this value in the method as isAvailable is only a short circuiting mechanism and the way to handle it depends on the caller.

tvernum commented:
Sorry - I did get part way through a review yesterday, but didn't finish it.

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:
```diff
@@ -541,49 +542,53 @@ public void onFailure(Exception e) {
-            executeAsyncWithOrigin(client.threadPool().getThreadContext(), SECURITY_ORIGIN,
-                client.prepareSearch(SECURITY_INDEX_NAME)
+        if (securityIndex.isAvailable() == false) {
+            listener.onResponse(Collections.emptyMap());
```
Ditto.

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/NativeRoleMappingStore.java:
```diff
@@ -220,32 +220,32 @@ public void onFailure(Exception e) {
-                client.prepareDelete(SECURITY_INDEX_NAME, SECURITY_GENERIC_TYPE, getIdForName(request.getName()))
+    private void innerDeleteMapping(DeleteRoleMappingRequest request, ActionListener<Boolean> listener) {
+        if (securityIndex.isAvailable() == false) {
+            listener.onResponse(false);
```
Likewise. I don't think we can claim that the mapping doesn't exist just because there are unavailable shards.

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java:
```diff
@@ -88,13 +88,15 @@ public NativePrivilegeStore(Settings settings, Client client, SecurityIndexManag
                               ActionListener<Collection<ApplicationPrivilegeDescriptor>> listener) {
-        if (applications != null && applications.size() == 1 && names != null && names.size() == 1) {
+        if (securityIndexManager.isAvailable() == false) {
+            listener.onResponse(Collections.emptyList());
```
Likewise.

jaymode commented:
@tvernum your concerns make sense and I think we can finally address some of these. The lack of a role retrieval result and authentication result prevented some of these changes previously. I will look at addressing your comments tomorrow. Do you mind finishing looking at this and leaving any other comments that you have?

jaymode commented:
I pushed 46c7b5ee6e53e33aa6f91c1cbd873810d319a400 to revert this on master and 65eeced9c076c69f963210a418fd0e1e2d2dc374 on 6.x. As discussed, I will open a new PR with fixes.

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:
```diff
@@ -673,30 +678,36 @@ private void findTokenFromRefreshToken(String refreshToken, ActionListener<Tuple
+            if (securityIndex.isAvailable() == false) {
+                logger.debug("security index is not available to find token from refresh token, retrying");
+                attemptCount.incrementAndGet();
+                findTokenFromRefreshToken(refreshToken, listener, attemptCount);
```
security index could have been deleted after generation of a refresh token?
